Sources/OpenAPIRuntime/Interface/HTTPBody.swift

```diff
@@ -0,0 +1,535 @@
+        )
+    }
+
+    public convenience init(
+        stream: AsyncThrowingStream<DataType, any Error>,
+        length: HTTPBody.Length
+    ) {
+        self.init(
+            sequence: .init(stream),
+            length: length,
+            iterationBehavior: .single
+        )
+    }
+
+    public convenience init(
+        stream: AsyncStream<DataType>,
+        length: HTTPBody.Length
+    ) {
+        self.init(
+            sequence: .init(stream),
+            length: length,
+            iterationBehavior: .single
+        )
+    }
+
+    public convenience init<S: AsyncSequence>(
+        sequence: S,
+        length: HTTPBody.Length,
+        iterationBehavior: IterationBehavior
+    ) where S.Element == DataType {
+        self.init(
+            sequence: .init(sequence),
+            length: length,
+            iterationBehavior: iterationBehavior
+        )
+    }
+}
+
+// MARK: - Consuming the body
+
+extension HTTPBody: AsyncSequence {
+    public typealias Element = DataType
+    public typealias AsyncIterator = Iterator
+    public func makeAsyncIterator() -> AsyncIterator {
+        if iterationBehavior == .single {
+            lock.lock()
+            defer {
+                lock.unlock()
+            }
+            guard !locked_iteratorCreated else {
+                fatalError(
+                    "OpenAPIRuntime.HTTPBody attempted to create a second iterator, but the underlying sequence is only safe to be iterated once."
+                )
+            }
+            locked_iteratorCreated = true
+        }
+        return sequence.makeAsyncIterator()
+    }
+}
+
+// MARK: - Transforming the body
+
+extension HTTPBody {
+
+    /// Creates a body where each chunk is transformed by the provided closure.
+    /// - Parameter transform: A mapping closure.
+    /// - Throws: If a known length was provided to this body at
+    /// creation time, the transform closure must not change the length of
+    /// each chunk.
+    public func mapChunks(
+        _ transform: @escaping @Sendable (Element) async -> Element
+    ) -> HTTPBody {
+        let validatedTransform: @Sendable (Element) async -> Element
+        switch length {
+        case .known:
+            validatedTransform = { element in
+                let transformedElement = await transform(element)
+                guard transformedElement.count == element.count else {
+                    fatalError(
+                        "OpenAPIRuntime.HTTPBody.mapChunks transform closure attempted to change the length of a chunk in a body which has a total length specified, this is not allowed."
+                    )
+                }
+                return transformedElement
+            }
+        case .unknown:
+            validatedTransform = transform
+        }
+        return HTTPBody(
+            sequence: map(validatedTransform),
+            length: length,
+            iterationBehavior: iterationBehavior
+        )
+    }
+}
```

Comment:
What's the use case for this method?

Reply:
https://github.com/apple/swift-openapi-runtime/pull/46#discussion_r1314593893

Reply:
All these use cases (streaming encryption/unencryption/zipping/unzipping) will not work as they will all change the chunk length. These are also stateful operations and the `@Sendable` requirement will make this needlessly hard. In addition they will usually buffer data and will need an explicit flush at the end which is not possible with this API. These use cases should just be modelled as `AsyncSequence` operators/transformers.

Reply:
Yeah I can be convinced, I'll remove the API, it's not pulling its weight.